File: src/main/java/eotree/EOProgram.java
package eotree;

import java.util.List;
import java.util.stream.Collectors;

/**
 * EBNF representation:
 * <code>
 * [ license ] [ metas ] { object eol }
 * </code>
 */
public class EOProgram extends EONode {
    public EOLicense   license;
    public EOMetas     metas;
    public List<EOBnd> bnds;

    public EOProgram(EOLicense license, EOMetas metas, List<EOBnd> bnds) {
        this.license = license;
        this.metas = metas;
        this.bnds = bnds;
    }

    @Override
    public String generateEO(int indent) {
        return license.generateEO(indent) +
               "\n" +
               metas.generateEO(indent) +
               "\n" +
               bnds.stream()
                       .map(bnd -> bnd.generateEO(indent))
                       .collect(Collectors.joining("\n")) + "\n";
    }
}

File: src/main/java/org/polystat/j2eo/eotree/EOProgram.kt
package org.polystat.j2eo.eotree

/**
 * EBNF representation:
 * `
 * [ license ] [ metas ] { object eol }
` *
 */
class EOProgram(var license: EOLicense, var metas: EOMetas, var bnds: List<EOBnd>) : EONode() {
    override fun generateEO(indent: Int): String =
        """${license.generateEO(indent)}
${metas.generateEO(indent)}

${bnds.joinToString("\n\n") { bnd: EOBnd -> bnd.generateEO(indent) }}"""

    override fun toString(): String = "[Program]"
}
